From 9098f6674b89eef4f0501cfb31d52ab4c009a489 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:11 -0400
Subject: [net] openvswitch: introduce rtnl ops stub

Message-id: <c1906253106645434a35d211842b715c1bdbca64.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94004
O-Subject: [RHEL7.1 net PATCH 38/50] openvswitch: introduce rtnl ops stub
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit 5b9e7e1607956e2454ccbd94ccf5631309ade054
Author: Jiri Pirko <jiri@resnulli.us>
Date:   Thu Jun 26 09:58:26 2014 +0200

    openvswitch: introduce rtnl ops stub

    This stub now allows userspace to see IFLA_INFO_KIND for ovs master and
    IFLA_INFO_SLAVE_KIND for slave.

    Signed-off-by: Jiri Pirko <jiri@resnulli.us>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/datapath.c b/net/openvswitch/datapath.c
index 775b51a..25e1b96 100644
--- a/net/openvswitch/datapath.c
+++ b/net/openvswitch/datapath.c
@@ -2055,10 +2055,14 @@ static int __init dp_init(void)
 
  pr_info("Open vSwitch switching datapath\n");
 
- err = ovs_flow_init();
+ err = ovs_internal_dev_rtnl_link_register();
  if (err)
   goto error;
 
+ err = ovs_flow_init();
+ if (err)
+  goto error_unreg_rtnl_link;
+
  err = ovs_vport_init();
  if (err)
   goto error_flow_exit;
@@ -2085,6 +2089,8 @@ error_vport_exit:
  ovs_vport_exit();
 error_flow_exit:
  ovs_flow_exit();
+error_unreg_rtnl_link:
+ ovs_internal_dev_rtnl_link_unregister();
 error:
  return err;
 }
@@ -2097,6 +2103,7 @@ static void dp_cleanup(void)
  rcu_barrier();
  ovs_vport_exit();
  ovs_flow_exit();
+ ovs_internal_dev_rtnl_link_unregister();
 }
 
 module_init(dp_init);
diff --git a/net/openvswitch/vport-internal_dev.c b/net/openvswitch/vport-internal_dev.c
index 789af92..295471a 100644
--- a/net/openvswitch/vport-internal_dev.c
+++ b/net/openvswitch/vport-internal_dev.c
@@ -26,6 +26,7 @@
 
 #include <net/dst.h>
 #include <net/xfrm.h>
+#include <net/rtnetlink.h>
 
 #include "datapath.h"
 #include "vport-internal_dev.h"
@@ -121,6 +122,10 @@ static const struct net_device_ops internal_dev_netdev_ops = {
  .ndo_get_stats64 = internal_dev_get_stats,
 };
 
+static struct rtnl_link_ops internal_dev_link_ops __read_mostly = {
+ .kind = "openvswitch",
+};
+
 static void do_setup(struct net_device *netdev)
 {
  ether_setup(netdev);
@@ -131,6 +136,7 @@ static void do_setup(struct net_device *netdev)
  netdev->priv_flags |= IFF_LIVE_ADDR_CHANGE;
  netdev->destructor = internal_dev_destructor;
  netdev->ethtool_ops = &internal_dev_ethtool_ops;
+ netdev->rtnl_link_ops = &internal_dev_link_ops;
  netdev->tx_queue_len = 0;
 
  netdev->features = NETIF_F_LLTX | NETIF_F_SG | NETIF_F_FRAGLIST |
@@ -248,3 +254,13 @@ struct vport *ovs_internal_dev_get_vport(struct net_device *netdev)
 
  return internal_dev_priv(netdev)->vport;
 }
+
+int ovs_internal_dev_rtnl_link_register(void)
+{
+ return rtnl_link_register(&internal_dev_link_ops);
+}
+
+void ovs_internal_dev_rtnl_link_unregister(void)
+{
+ rtnl_link_unregister(&internal_dev_link_ops);
+}
diff --git a/net/openvswitch/vport-internal_dev.h b/net/openvswitch/vport-internal_dev.h
index 9a7d30e..1b179a1 100644
--- a/net/openvswitch/vport-internal_dev.h
+++ b/net/openvswitch/vport-internal_dev.h
@@ -24,5 +24,7 @@
 
 int ovs_is_internal_dev(const struct net_device *);
 struct vport *ovs_internal_dev_get_vport(struct net_device *);
+int ovs_internal_dev_rtnl_link_register(void);
+void ovs_internal_dev_rtnl_link_unregister(void);
 
 #endif /* vport-internal_dev.h */
-- 
1.7.1